src/salus.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2023 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Salus vendor exception.
+
+use crate::error::*;
+use crate::vendor::*;
+use crate::SbiFunction;
+
+const EXT_SALUS_TEST: u64 = 0x09FFFFFF;
+
+pub type SalusSbiMessage = VendorSbiMessage<SalusExtension>;
+
+pub enum SalusExtension {
+    SalusTest(SalusTestFunction),
+}
+
+impl VendorExtension for SalusExtension {
+    fn from_regs(args: &[u64]) -> Result<Self> {
+        use SalusExtension::*;
+        match args[7] {
+            EXT_SALUS_TEST => SalusTestFunction::from_regs(args).map(SalusTest),
+            _ => Err(Error::NotSupported),
+        }
+    }
+
+    fn eid(&self) -> u64 {
+        use SalusExtension::*;
+        match self {
+            SalusTest(_) => EXT_SALUS_TEST,
+        }
+    }
+}
+
+impl SbiFunction for SalusExtension {
+    fn a0(&self) -> u64 {
+        use SalusExtension::*;
+        match self {
+            SalusTest(m) => m.a0(),
+        }
+    }
+
+    fn a1(&self) -> u64 {
+        use SalusExtension::*;
+        match self {
+            SalusTest(m) => m.a1(),
+        }
+    }
+
+    fn a2(&self) -> u64 {
+        use SalusExtension::*;
+        match self {
+            SalusTest(m) => m.a2(),
+        }
+    }
+
+    fn a3(&self) -> u64 {
+        use SalusExtension::*;
+        match self {
+            SalusTest(m) => m.a3(),
+        }
+    }
+
+    fn a4(&self) -> u64 {
+        use SalusExtension::*;
+        match self {
+            SalusTest(m) => m.a4(),
+        }
+    }
+
+    fn a5(&self) -> u64 {
+        use SalusExtension::*;
+        match self {
+            SalusTest(m) => m.a5(),
+        }
+    }
+
+    fn a6(&self) -> u64 {
```

Comment:
Increasingly keen on turning these into a macro!.

Reply:
Agreed. This will either require a `enum_dispatch` or `paste` as suggested. I found doing this being a distraction of what this patch should do -- made the code slightly harder to follow, so postponed to future work.